### ISSUE TITLE
Cut down on calls to VecGetOwnershipRange

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#445]
This shows up surprisingly often in callgrind profiling: it is about 5% of our total run time. We call it a lot, apparently, in  `IBTK::get_values_for_interpolation` and `IBTK::FEDataInterpolation::collectDataForInterpolation`. We should find a way to cut down on these calls: we call it roughly one million times per time step which is silly.

This function is dead simple: it just [reads two integers](https://www.mcs.anl.gov/petsc/petsc-current/src/vec/vec/interface/vector.c.html#VecGetOwnershipRange). I don't know why we are spending so many cycles on it.

**Comment by knepley** [comment]
That sounds very suspicious. Are you sure callgrind is not screwing up? That function should be inlined in optimized mode anyway.

**Comment by knepley** [comment]
Crap. The linked code is misleading because Barry has stripped out the checking code. If you call any function that often, the checking code can become onerous. Can you try two things for me:
1) Change the first line to "PetscFunctionBeginHot" and rebuild. This will turn off stack building here, which is unnecessary anyway.
2) Run using -check_pointer_intensity 0, which turns off those pointer checks
3) The checks probably prevent inling as well. So if the above do not work, we can make a stripped down internal version that will definitely be inlined.

**Comment by drwells** [comment]
> Are you sure callgrind is not screwing up?

"Its not paranoia if they are actually out to get you"

More seriously: unless I am mistaken this function is not marked as inline: I see 
```c
PETSC_EXTERN PetscErrorCode VecGetOwnershipRange(Vec,PetscInt*,PetscInt*);
```
in `petscvec.h` (PETSc 3.10.0) so I don't see how it could possibly be inlined.

**Comment by drwells** [comment]
I did all the profiling with my 'system' copy of PETSc on a personal machine: I'll have to recompile a lot of stuff to try that patch so I cannot test this immediately. I'll try to get to it later this week.

**Comment by knepley** [comment]
Very short functions, if the only thing that happened was setting the values, will be automatically inlined by an optimizing compiler without the PETSC_INLINE directive.

**Comment by drwells** [comment]
I don't see how this can happen without requesting link time optimization, in the current case; the implementation is not in a header (its in `src/vec/vec/interface/vector.c`).

**Comment by knepley** [comment]
Ah, yes. Only in the same compilation unit. Okay, we could move it inside the header if those optimizations I sent do not work.

**Comment by drwells** [comment]
I think I figured this out from inside IBAMR. We are calling this function at every single quadrature point, which is really unnecessary, but we can avoid it. I'll have a patch later today or tomorrow.

**Comment by boyceg** [comment]
`get_values_for_interpolation` is ripe for optimization. We should be able to cache & reuse the mapping from DOFs to local indices.

**Comment by drwells** [comment]
For non-ghost indices the mapping is really simple: just subtract an offset.

**Comment by boyceg** [comment]
Sure, but you have to check if you are in the local range or not — and isn’t that the function that is showing up prominently in callgrind?

**Comment by drwells** [comment]
Yes, but if we poke `libMesh` in just the right way it will write down those values and stop calling `VecGetOwnershipRange` every time we need to do a range check.

**Comment by drwells** [comment]
Fixed by #457.
